server/raft.go

```diff
@@ -2311,34 +2312,40 @@ func (ae *appendEntry) encode(b []byte) ([]byte, error) {
 
 	var elen int
 	for _, e := range ae.entries {
+		if len(e.Data) > math.MaxUint32 {
+			return nil, errBadAppendEntry
+		}
 		elen += len(e.Data) + 1 + 4 // 1 is type, 4 is for size.
 	}
-	tlen := appendEntryBaseLen + elen + 1
+	// Uvarint for lterm can be a maximum 10 bytes for a uint64.
+	var _lterm [10]byte
+	lterm := _lterm[:binary.PutUvarint(_lterm[:], ae.lterm)]
+	tlen := appendEntryBaseLen + elen + len(lterm)
 
 	var buf []byte
 	if cap(b) >= tlen {
-		buf = b[:tlen]
+		buf = b[:idLen]
 	} else {
-		buf = make([]byte, tlen)
+		buf = make([]byte, idLen, tlen)
 	}
 
 	var le = binary.LittleEndian
 	copy(buf[:idLen], ae.leader)
-	le.PutUint64(buf[8:], ae.term)
-	le.PutUint64(buf[16:], ae.commit)
-	le.PutUint64(buf[24:], ae.pterm)
-	le.PutUint64(buf[32:], ae.pindex)
-	le.PutUint16(buf[40:], uint16(len(ae.entries)))
-	wi := 42
+	buf = le.AppendUint64(buf, ae.term)
```

Comment:
These used to be very slow compared to Put. I think @neilalexander might have measured lately but when I tested was very slow comparatively.

Reply:
It's dependent on capacity. In this case, capacity is preallocated therefore the appends are very fast.